File: frontend/src/types/review.ts
/**
 * Review System Type Definitions
 *
 * CRITICAL: These types MUST match backend models exactly:
 * - app/models/review_request.py
 * - app/models/review_correction.py
 * - app/models/enums.py (ReviewRequestStatus, ReviewPriority)
 */

/**
 * Review request status lifecycle
 * pending → assigned → in_review → completed
 *             ↓            ↓
 *         cancelled    escalated
 */
export enum ReviewRequestStatus {
  PENDING = "pending",
  ASSIGNED = "assigned",
  IN_REVIEW = "in_review",
  COMPLETED = "completed",
  CANCELLED = "cancelled",
  ESCALATED = "escalated",
}

/**
 * Review priority levels
 */
export enum ReviewPriority {
  CRITICAL = "critical",
  HIGH = "high",
  NORMAL = "normal",
  LOW = "low",
}

/**
 * Correction types for tracking changes
 * MUST match backend: app/models/enums.py CorrectionType
 */
export enum CorrectionType {
  VALUE_CHANGE = "value_change",
  FIELD_ADDITION = "field_addition",
  FIELD_REMOVAL = "field_removal",
  TYPE_CORRECTION = "type_correction",
}

/**
 * Review request model (matches backend ReviewRequest)
 */
export interface ReviewRequest {
  id: string;
  tenant_id: string;
  extraction_job_id: string;
  document_id?: string; // Populated via join

  // Routing information
  trigger_reason: string;
  confidence_score: number | null;
  priority: ReviewPriority;

  // Assignment information
  status: ReviewRequestStatus;
  assigned_to_user_id: string | null;
  assigned_to_user_name?: string; // Populated via join
  assigned_at: string | null;

  // SLA tracking
  sla_minutes: number;
  sla_deadline: string | null;
  escalated: boolean;
  escalated_at: string | null;
  escalated_to_user_id: string | null;
  escalated_to_user_name?: string; // Populated via join

  // Completion tracking
  completed_at: string | null;
  review_time_minutes: number | null;

  // Metadata
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;

  // Joined data (from API response)
  extraction_job?: {
    id: string;
    document_id: string;
    schema_name?: string;
    status: string;
  };
  document?: {
    id: string;
    filename: string;
    page_count: number;
  };
  extracted_data?: Record<string, unknown>; // Current extraction result
}

/**
 * Review correction model (matches backend ReviewCorrection)
 */
export interface ReviewCorrection {
  id?: string; // Optional for new corrections
  review_request_id?: string;
  extraction_result_id: string;

  // Field information
  field_path: string; // JSONPath to field (e.g., "invoice.total")
  original_value: unknown;
  corrected_value: unknown;

  // Correction metadata
  correction_type: CorrectionType;
  correction_notes: string | null;
  reviewer_confidence: number | null; // 0.0-1.0

  // Timestamps
  created_at?: string;
}

/**
 * Queue filter parameters
 */
export interface QueueFilters {
  status?: ReviewRequestStatus | ReviewRequestStatus[];
  priority?: ReviewPriority | ReviewPriority[];
  assigned_to_me?: boolean;
  unassigned?: boolean;
  escalated?: boolean;
  sort_by?: "priority" | "age" | "confidence" | "sla_deadline";
  sort_order?: "asc" | "desc";
  page?: number;
  page_size?: number;
}

/**
 * Review queue response
 */
export interface ReviewQueueResponse {
  items: ReviewRequest[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

/**
 * Review metrics for dashboard
 */
export interface ReviewMetrics {
  pending_count: number;
  assigned_count: number;
  in_review_count: number;
  completed_today: number;
  avg_review_time_minutes: number;
  sla_breach_count: number;
  accuracy_rate: number; // % of reviews with corrections
  my_queue_count: number;
  escalated_count: number;
}

/**
 * Submit review request
 */
export interface SubmitReviewRequest {
  corrections: ReviewCorrection[];
  review_notes?: string;
  action: "approve" | "reject" | "escalate";
}

/**
 * Assign review request
 */
export interface AssignReviewRequest {
  user_id: string;
}

/**
 * Type guards for runtime validation
 */
export function isReviewRequestStatus(value: string): value is ReviewRequestStatus {
  return Object.values(ReviewRequestStatus).includes(value as ReviewRequestStatus);
}

export function isReviewPriority(value: string): value is ReviewPriority {
  return Object.values(ReviewPriority).includes(value as ReviewPriority);
}

/**
 * Helper functions for display
 */
export function getReviewStatusLabel(status: ReviewRequestStatus): string {
  const labels: Record<ReviewRequestStatus, string> = {
    [ReviewRequestStatus.PENDING]: "Pending",
    [ReviewRequestStatus.ASSIGNED]: "Assigned",
    [ReviewRequestStatus.IN_REVIEW]: "In Review",
    [ReviewRequestStatus.COMPLETED]: "Completed",
    [ReviewRequestStatus.CANCELLED]: "Cancelled",
    [ReviewRequestStatus.ESCALATED]: "Escalated",
  };
  return labels[status];
}

export function getReviewStatusColor(status: ReviewRequestStatus): string {
  const colors: Record<ReviewRequestStatus, string> = {
    [ReviewRequestStatus.PENDING]: "text-gray-600",
    [ReviewRequestStatus.ASSIGNED]: "text-blue-600",
    [ReviewRequestStatus.IN_REVIEW]: "text-yellow-600",
    [ReviewRequestStatus.COMPLETED]: "text-green-600",
    [ReviewRequestStatus.CANCELLED]: "text-gray-400",
    [ReviewRequestStatus.ESCALATED]: "text-red-600",
  };
  return colors[status];
}

export function getReviewPriorityLabel(priority: ReviewPriority): string {
  const labels: Record<ReviewPriority, string> = {
    [ReviewPriority.CRITICAL]: "Critical",
    [ReviewPriority.HIGH]: "High",
    [ReviewPriority.NORMAL]: "Normal",
    [ReviewPriority.LOW]: "Low",
  };
  return labels[priority];
}

export function getReviewPriorityColor(priority: ReviewPriority): string {
  const colors: Record<ReviewPriority, string> = {
    [ReviewPriority.CRITICAL]: "text-red-600",
    [ReviewPriority.HIGH]: "text-orange-600",
    [ReviewPriority.NORMAL]: "text-blue-600",
    [ReviewPriority.LOW]: "text-gray-600",
  };
  return colors[priority];
}

export function getReviewPriorityIcon(priority: ReviewPriority): string {
  const icons: Record<ReviewPriority, string> = {
    [ReviewPriority.CRITICAL]: "⚡",
    [ReviewPriority.HIGH]: "🔶",
    [ReviewPriority.NORMAL]: "🟢",
    [ReviewPriority.LOW]: "⚪",
  };
  return icons[priority];
}

/**
 * Calculate confidence level from score
 */
export function getConfidenceLevel(
  score: number | null
): "high" | "medium" | "low" | "unknown" {
  if (score === null) return "unknown";
  if (score >= 0.8) return "high";
  if (score >= 0.5) return "medium";
  return "low";
}

export function getConfidenceBadgeProps(score: number | null): {
  label: string;
  color: string;
  icon: string;
} {
  const level = getConfidenceLevel(score);

  const props = {
    high: { label: "High Confidence", color: "text-green-600", icon: "✓" },
    medium: { label: "Medium Confidence", color: "text-orange-600", icon: "⚠️" },
    low: { label: "Low Confidence", color: "text-red-600", icon: "❌" },
    unknown: { label: "Unknown", color: "text-gray-600", icon: "?" },
  };

  return props[level];
}

/**
 * Calculate SLA status
 */
export function getSLAStatus(
  slaDeadline: string | null,
  status: ReviewRequestStatus
): "ok" | "warning" | "breached" | "completed" {
  if (status === ReviewRequestStatus.COMPLETED || status === ReviewRequestStatus.CANCELLED) {
    return "completed";
  }

  if (!slaDeadline) return "ok";

  const deadline = new Date(slaDeadline);
  const now = new Date();
  const hoursRemaining = (deadline.getTime() - now.getTime()) / (1000 * 60 * 60);

  if (hoursRemaining < 0) return "breached";
  if (hoursRemaining < 1) return "warning";
  return "ok";
}

export function getSLAStatusColor(slaStatus: "ok" | "warning" | "breached" | "completed"): string {
  const colors = {
    ok: "text-green-600",
    warning: "text-orange-600",
    breached: "text-red-600",
    completed: "text-gray-400",
  };
  return colors[slaStatus];
}

File: app/schemas/review.py
"""Review-related Pydantic schemas for HITL system."""

from datetime import datetime
from typing import Any, Optional, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ReviewRequestStatus, ReviewPriority, CorrectionType


# Request/Response Schemas

class CorrectionCreate(BaseModel):
    """Schema for creating a correction."""

    extraction_result_id: UUID = Field(..., description="ID of extraction result being corrected")
    field_path: str = Field(..., description="JSONPath to field (e.g., 'invoice.total')")
    original_value: Optional[Any] = Field(None, description="Original AI-extracted value")
    corrected_value: Optional[Any] = Field(None, description="Human-corrected value")
    correction_type: CorrectionType = Field(..., description="Type of correction")
    correction_notes: Optional[str] = Field(None, description="Explanation for correction")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "extraction_result_id": "123e4567-e89b-12d3-a456-426614174000",
                "field_path": "invoice.total",
                "original_value": "100.00",
                "corrected_value": "150.00",
                "correction_type": "value_change",
                "correction_notes": "Handwritten total was misread by AI"
            }
        }
    )


class CorrectionResponse(BaseModel):
    """Schema for correction response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    review_request_id: UUID
    extraction_result_id: UUID
    field_path: str
    original_value: Optional[Any]
    corrected_value: Optional[Any]
    correction_type: str
    correction_notes: Optional[str]
    corrected_by_user_id: Optional[UUID]
    created_at: datetime


class ReviewRequestCreate(BaseModel):
    """Schema for manual review request creation."""

    extraction_job_id: UUID = Field(..., description="Extraction job ID to review")
    priority: Optional[ReviewPriority] = Field(
        ReviewPriority.NORMAL,
        description="Review priority (affects SLA)"
    )
    trigger_reason: Optional[str] = Field(
        "manual_request",
        description="Reason for review request"
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "extraction_job_id": "123e4567-e89b-12d3-a456-426614174000",
                "priority": "high",
                "trigger_reason": "manual_request"
            }
        }
    )


class ReviewRequestResponse(BaseModel):
    """Schema for review request response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    extraction_job_id: UUID
    assigned_to_user_id: Optional[UUID]
    status: str
    priority: str
    confidence_score: float
    trigger_reason: Optional[str]
    review_notes: Optional[str]
    conductor_task_id: Optional[str]
    conductor_workflow_id: Optional[str]
    sla_deadline: datetime
    created_at: datetime
    assigned_at: Optional[datetime]
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    updated_at: datetime


class ReviewAssignRequest(BaseModel):
    """Schema for assigning a review."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "user_id": "123e4567-e89b-12d3-a456-426614174000"
            }
        }
    )

    user_id: UUID = Field(..., description="User ID to assign review to")


class ReviewEscalateRequest(BaseModel):
    """Schema for escalating a review."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "reason": "Document requires specialist expertise in technical terminology"
            }
        }
    )

    reason: str = Field(..., min_length=10, max_length=1000, description="Reason for escalation")


class ReviewSubmitRequest(BaseModel):
    """Schema for submitting review corrections."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "corrections": [
                    {
                        "extraction_result_id": "123e4567-e89b-12d3-a456-426614174000",
                        "field_path": "invoice.total",
                        "original_value": "100.00",
                        "corrected_value": "150.00",
                        "correction_type": "value_change",
                        "correction_notes": "Handwritten total was misread"
                    }
                ],
                "review_notes": "Document quality was poor, several fields required correction"
            }
        }
    )

    corrections: List[CorrectionCreate] = Field(
        ...,
        description="List of corrections made"
    )
    review_notes: Optional[str] = Field(
        None,
        description="General notes about the review"
    )


class ReviewQueueFilter(BaseModel):
    """Schema for filtering review queue."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "status": "pending",
                "priority": "high",
                "page": 1,
                "page_size": 50
            }
        }
    )

    status: Optional[ReviewRequestStatus] = Field(None, description="Filter by status")
    priority: Optional[ReviewPriority] = Field(None, description="Filter by priority")
    assigned_to_user_id: Optional[UUID] = Field(None, description="Filter by assignee")
    page: int = Field(1, ge=1, description="Page number")
    page_size: int = Field(50, ge=1, le=100, description="Items per page")


class ReviewQueueResponse(BaseModel):
    """Schema for review queue response."""

    model_config = ConfigDict(from_attributes=True)

    items: List[ReviewRequestResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class ReviewMetricsResponse(BaseModel):
    """Schema for review metrics response."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "total_reviews": 1000,
                "completed_reviews": 950,
                "pending_reviews": 50,
                "average_review_time_minutes": 15.5,
                "sla_breach_count": 5,
                "accuracy_by_schema": {
                    "invoice_schema": 0.95,
                    "receipt_schema": 0.88
                },
                "most_corrected_fields": [
                    {"field_path": "invoice.total", "correction_count": 45},
                    {"field_path": "invoice.date", "correction_count": 32}
                ]
            }
        }
    )

    total_reviews: int
    completed_reviews: int
    pending_reviews: int
    average_review_time_minutes: Optional[float]
    sla_breach_count: int
    accuracy_by_schema: dict[str, float]
    most_corrected_fields: List[dict[str, Any]]
